cpp/src/gandiva/precompiled/decimal_wrapper.cc

```diff
@@ -231,4 +231,131 @@ void castDECIMAL_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_p
+                                           boolean x_isvalid, int64_t seed,
+                                           boolean seed_isvalid) {
+  if (!x_isvalid) {
+    return seed;
+  }
+  return hash64_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16, seed);
+}
+
+FORCE_INLINE
+int32_t hash32AsDouble_decimal128_internal(int64_t x_high, uint64_t x_low,
+                                           int32_t x_precision, int32_t x_scale,
+                                           boolean x_isvalid) {
+  return x_isvalid
+             ? hash32_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16, 0)
+             : 0;
+}
+
+FORCE_INLINE
+int64_t hash64AsDouble_decimal128_internal(int64_t x_high, uint64_t x_low,
+                                           int32_t x_precision, int32_t x_scale,
+                                           boolean x_isvalid) {
+  return x_isvalid
+             ? hash64_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16, 0)
+             : 0;
+}
+
+FORCE_INLINE
+int32_t hash32AsDoubleWithSeed_decimal128_internal(int64_t x_high, uint64_t x_low,
+                                                   int32_t x_precision, int32_t x_scale,
+                                                   boolean x_isvalid, int32_t seed,
+                                                   boolean seed_isvalid) {
+  if (!x_isvalid) {
+    return seed;
+  }
+  return hash32_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16, seed);
+}
+
+FORCE_INLINE
+int64_t hash64AsDoubleWithSeed_decimal128_internal(int64_t x_high, uint64_t x_low,
+                                                   int32_t x_precision, int32_t x_scale,
+                                                   boolean x_isvalid, int64_t seed,
+                                                   boolean seed_isvalid) {
+  if (!x_isvalid) {
+    return seed;
+  }
+  return hash64_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16, seed);
+}
+
+FORCE_INLINE
+boolean isnull_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_precision,
+                                   int32_t x_scale, boolean x_isvalid) {
+  return !x_isvalid;
+}
+
+FORCE_INLINE
+boolean isnotnull_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_precision,
+                                      int32_t x_scale, boolean x_isvalid) {
+  return x_isvalid;
+}
+
+FORCE_INLINE
+boolean isnumeric_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_precision,
+                                      int32_t x_scale, boolean x_isvalid) {
+  return x_isvalid;
+}
+
+FORCE_INLINE
+boolean is_not_distinct_from_decimal128_decimal128_internal(
+    int64_t x_high, uint64_t x_low, int32_t x_precision, int32_t x_scale,
+    boolean x_isvalid, int64_t y_high, uint64_t y_low, int32_t y_precision,
+    int32_t y_scale, boolean y_isvalid) {
+  if (x_isvalid != y_isvalid) {
+    return false;
+  }
+  if (!x_isvalid) {
+    return true;
```

Comment:
sorry did not notice before 
should we not return false here? one null is never equal to another null..

Reply:
please ignore looks like this is what sql standard recommends.

https://en.wikipedia.org/wiki/Null_(SQL)#When_two_nulls_are_equal:_grouping,_sorting,_and_some_set_operations